### ISSUE TITLE
Problems encountered in fine-tuning models

### ISSUE THREAD

**Issue by Annie1900** [magic-research/PLLaVA#13]
I'm trying to fine-tune the model with a specific data set and expect the model to output in a specified format. The amount of data used was 1000 generic data +1500 formatted data, and with 50 epochs fine-tuned, the loss was reduced from 3 to 0.9 and kept convergent. However, the output of the model is not in the specified format but a whole paragraph of text, but the content expression is similar to our data. Why does this happen? Is it related to the use of lora? What should I do? Thank you so much!!
打扰啦，劳动节快乐，我正在尝试以特定的数据集微调模型，期望模型按指定格式输出。使用的数据量是1000个通用数据+1500个特定格式的数据，微调了50个epoch，loss从3降低到了0.9并且保持收敛。然而，模型的输出并没有按指定格式而是一整段文本，但内容表达上与我们的数据相似，请问为什么会出现这样的情况？是否与使用了lora有关？我应该怎么做？非常感谢！！！

**Comment by Annie1900** [comment]
The first is the data format I used to fine-tune and expect to get, and the second is the inference result of the fine-tuned model
`time:Daytime\nground:Indoors, there is a wooden dining table and chairs.\nelement:Laptop, pen, notebook, desk lamp, plant.\nangle:Slanting\nheight:Flat shot\nfield:Foreground\ncamera:Fixed, Mobile.\nrhythm:Slow motion\nstyle:Real scene\ntopic_chose:Work, Life\natmosphere:Focused, calm\ndesp_people:A man and a woman are sitting across from each other at a desk, with a laptop on the side. The man is pointing at the documents on the desk and explaining them to the other person.\nhighlight:4\naudio:No\nsubtitle:No\nwatermark:None`
![image](https://github.com/magic-research/PLLaVA/assets/73414003/7581ccaf-59bf-414c-8470-816740828fa5)

**Comment by Annie1900** [comment]
The first is the data format I used to fine-tune and expect to get, and the second is the inference result of the fine-tuned model
"time:Daytime\nground:Indoors, there is a wooden dining table and chairs.\nelement:Laptop, pen, notebook, desk lamp, plant.\nangle:Slanting\nheight:Flat shot\nfield:Foreground\ncamera:Fixed, Mobile.\nrhythm:Slow motion\nstyle:Real scene\ntopic_chose:Work, Life\natmosphere:Focused, calm\ndesp_people:A man and a woman are sitting across from each other at a desk, with a laptop on the side. The man is pointing at the documents on the desk and explaining them to the other person.\nhighlight:4\naudio:No\nsubtitle:No\nwatermark:None"
![image](https://github.com/magic-research/PLLaVA/assets/73414003/cc6b9252-4650-4d27-ab9a-15490cbd8b90)

**Comment by Annie1900** [comment]
In terms of weights, I notice that output_dir has three folders: ckpt_epoch, pretrained_epoch, and pretrained_step. I noticed that the format of combining these three folders is the same as pllava_7b you provided, so I combined the contents of these three folders to form a new folder pllava_51 as weight_dir. My demo is as follows:
在权重上，我注意到output_dir有三个文件夹，分别是ckpt_epoch、pretrained_epoch、pretrained_step。我注意到将这三个文件夹合并起来后的格式与您提供的pllava_7b相同，所以我将这三个文件夹的内容合并形成新的文件夹pllava_51作为weight_dir，我的demo如下：
![image](https://github.com/magic-research/PLLaVA/assets/73414003/e58fb607-4d55-4264-8570-a668da0bcc84)

**Comment by Annie1900** [comment]
My model setup is as follows, I really admire your work and wonder why I can't fine-tune the model output in the specified format, thank you very much.
`  "train_corpus": "data_finance",
  "train_file": [
    [
      "dataset/pllava/finance_tune_pllava_1_429.json",
      "dataset/llava_finance",
      "video"
    ]
  ],
  "test_file": {},
  "test_types": [],
  "num_workers": 4,
  "save_steps": 10000,
  "ckpt_steps": 10000,
  "stop_key": null,
  "deepspeed": false,
  "num_frames": 16,
  "num_frames_test": 1,
  "batch_size": 8,
  "gradient_accumulation_steps": 16,
  "max_txt_l": 512,
  "max_train_steps": null,
  "pre_text": false,
  "gradient_checkpointing": true,
  "inputs": {
    "image_res": 336,
    "video_input": {
      "num_frames": 16,
      "sample_type": "rand",
      "num_frames_test": 1,
      "sample_type_test": "middle",
      "random_aug": false
    },
    "max_txt_l": {
      "image": 512,
      "video": 512
    },
    "batch_size": {
      "image": 8,
      "video": 8
    },
    "batch_size_test": {
      "image": 8,
      "video": 8
    }
  },
  "model": {
    "repo_id": "llava-hf/llava-v1.6-vicuna-7b-hf",
    "pretrained_path": "checkpoint/ermu2001/pllava-7b",
    "load_from_origin": false,
    "origin_vision": "",
    "origin_llm": "",
    "vision_encoder": {
      "name": "vit_l14"
    },
    "torch_dtype": "bfloat16",
    "freeze_projector": false,
    "freeze_lm": false,
    "freeze_vision_tower": true,
    "lora_target_modules": [
      "q_proj",
      "v_proj"
    ],
    "use_lora": true,
    "lora_r": 128,
    "lora_alpha": 32,
    "lora_dropout": 0.05,
    "num_frames": 16,
    "pooling_method": "avg",
    "use_pooling": true,
    "frame_shape": [
      24,
      24
    ],
    "pooling_shape": [
      16,
      12,
      12
    ]
  },
  "preprocess": {
    "system": "",
    "mm_alone": false,
    "random_shuffle": false,
    "add_second_msg": false,
    "roles": [
      "USER:",
      "ASSISTANT:"
    ],
    "end_signal": [
      " ",
      "</s>"
    ],
    "begin_signal": "",
    "dataset_image_placeholder": "<Image></Image>",
    "dataset_video_placeholder": "<Video></Video>",
    "image_token_index": 32000,
    "max_txt_l": 512,
    "ignore_index": -100,
    "center_pad": false,
    "longest_edge": 762,
    "shortest_edge": 336,
    "clip_transform": false,
    "num_frames": 16
  },
  "optimizer": {
    "opt": "adamW",
    "lr": 0.0002,
    "opt_betas": [
      0.9,
      0.999
    ],
    "weight_decay": 0.02,
    "max_grad_norm": -1,
    "different_lr": {
      "enable": false,
      "module_names": [],
      "lr": 0.001
    }
  },
  "scheduler": {
    "is_videochat2_custom": false,
    "sched": "cosine",
    "epochs": 50,
    "warmup_ratio": 0.2,
    "min_lr_multi": 0.25
  },
  "evaluate": false,
  "deep_fusion": false,
  "evaluation": {
    "eval_frame_ensemble": "concat",
    "eval_x_only": false,
    "k_test": 128,
    "eval_offload": true
  },
  "fp16": true,
  "wandb": {
    "enable": false,
    "entity": "user",
    "project": "videochat2"
  },
  "dist_url": "env://",
  "device": "cuda",
  "mode": "it",
  "output_dir": "./pllava_video_outputs/test_train_7b_reconstruct_430_0",
  "resume": false,
  "debug": false,
  "log_freq": 5,
  "metric_window_size": 10,
  "seed": 42,
  "report_to": "tensorboard",
  "save_latest": true,
  "auto_resume": true,
  "pretrained_path": "",
  "rank": 0,
  "world_size": 4,
  "gpu": 0,
  "distributed": true,
  "dist_backend": "nccl"
}`

**Comment by Annie1900** [comment]
> Glad that you got our training pipe running, congrats!
> 
> Not sure whether this scale of a training data would be useful for finetuning on.
> 
> But regarding the formatting of the response, I think it would be better to try prompting the model. I tried prompting the 34b model, and it actually done a great job... (Even gone beyond my expectation😂)
> 
> ![image](https://private-user-images.githubusercontent.com/55656210/327039274-492489aa-ebf6-4e9d-926d-dccb15e7397e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NjYzMTEsIm5iZiI6MTcxNDU2NjAxMSwicGF0aCI6Ii81NTY1NjIxMC8zMjcwMzkyNzQtNDkyNDg5YWEtZWJmNi00ZTlkLTkyNmQtZGNjYjE1ZTczOTdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDEyMjAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0ZDI0NDk2OTI2MTExNzJkN2UzOGIxMjUxZTc5MDgwYzAyMDJlM2Q5YWU5YTJhNmY2NmIzOTRlZGIzYTZjNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cFTy5B_XJxVgaDf5zYT1iYsVNiVNmUW7HpR1Vt6koaI)
> 
> You can try prompting the model with this query:
> 
> You are to answer my question in a formated way, For example, you response for another video should have the expected format:
> 
> ```
> time:Daytime
> ground:Indoors, there is a wooden dining table and chairs.
> element:Laptop, pen, notebook, desk lamp, plant.
> angle:Slanting
> height:Flat shot
> field:Foreground
> camera:Fixed, Mobile.
> rhythm:Slow motion
> style:Real scene
> topic_chose:Work, Life
> atmosphere:Focused, calm
> desp_people:A man and a woman are sitting across from each other at a desk, with a laptop on the side. The man is pointing at the documents on the desk and explaining them to the other person.
> highlight:4
> audio:No
> subtitle:No
> watermark:None
> ```
> 
> Do not focus on the contents in this example response, instead, only take the response format and answer my question: Provide all dimensions for this video, details including: "time", "ground" , etc.



> Glad that you got our training pipe running, congrats!
> 
> Not sure whether this scale of a training data would be useful for finetuning on.
> 
> But regarding the formatting of the response, I think it would be better to try prompting the model. I tried prompting the 34b model, and it actually done a great job... (Even gone beyond my expectation😂)
> 
> ![image](https://private-user-images.githubusercontent.com/55656210/327039274-492489aa-ebf6-4e9d-926d-dccb15e7397e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NjYzMTEsIm5iZiI6MTcxNDU2NjAxMSwicGF0aCI6Ii81NTY1NjIxMC8zMjcwMzkyNzQtNDkyNDg5YWEtZWJmNi00ZTlkLTkyNmQtZGNjYjE1ZTczOTdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDEyMjAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0ZDI0NDk2OTI2MTExNzJkN2UzOGIxMjUxZTc5MDgwYzAyMDJlM2Q5YWU5YTJhNmY2NmIzOTRlZGIzYTZjNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cFTy5B_XJxVgaDf5zYT1iYsVNiVNmUW7HpR1Vt6koaI)
> 
> You can try prompting the model with this query:
> 
> You are to answer my question in a formated way, For example, you response for another video should have the expected format:
> 
> ```
> time:Daytime
> ground:Indoors, there is a wooden dining table and chairs.
> element:Laptop, pen, notebook, desk lamp, plant.
> angle:Slanting
> height:Flat shot
> field:Foreground
> camera:Fixed, Mobile.
> rhythm:Slow motion
> style:Real scene
> topic_chose:Work, Life
> atmosphere:Focused, calm
> desp_people:A man and a woman are sitting across from each other at a desk, with a laptop on the side. The man is pointing at the documents on the desk and explaining them to the other person.
> highlight:4
> audio:No
> subtitle:No
> watermark:None
> ```
> 
> Do not focus on the contents in this example response, instead, only take the response format and answer my question: Provide all dimensions for this video, details including: "time", "ground" , etc.

Wow, that's very kind of you, I will try this method, it looks like it will help us quite a bit, thank you very much!
